tests/providers/test_company.py

```diff
@@ -322,3 +327,30 @@ def test_bs(self, faker, num_samples):
             bs_words = bs.split()
             assert isinstance(bs, str)
             assert bs_words[0] in RuRuCompanyProvider.bsWords[0]
+
+
+class TestItIt:
+    """Test it_IT company provider methods"""
+
+    vat_regex = re.compile(r"^IT\d{7}(0\d{2}|100|120|121|888|999)\d$", flags=re.ASCII)
+
+    def test_company_vat(self, faker, num_samples):
+        for _ in range(num_samples):
+            company_vat = faker.company_vat()
+            assert self.vat_regex.match(company_vat)
+
+    @pytest.mark.parametrize("value, expected", (
+        (100, "100"),
+        (101, "120"),
+        (102, "121"),
+        (103, "888"),
+        (104, "999"),
+    ))
+    def test_company_vat_mock(self, value, expected):
+        # this test allows to get full code coverage for company_vat fixing the internal state of the random generator
+        fake = ItItCompanyProvider(generator=Generator())
```

Comment:
`Generator` and `Factory` can be considered legacy and for internal use only. Personally, I want to see them and providers redesigned to reduce the amount of magic (and abuse) going on under the hood, but it will be difficult for us to proceed if the test suite uses them everywhere.

On that note, there is no need to instantiate a `Generator` object. Remove the import, and use the `faker` pytest fixture instead like so:

```python
@pytest.mark.parametrize("value, expected", (
    # ...
))
def test_company_vat_special_cases(self, faker, value, expected):
    # ...
    provider = ItItCompanyProvider(generator=faker)
    # ...
```

Reply:
In fact I was trying to using directly a `Faker` instance, but I had trouble to properly mock the internal random state.
I'll update the PR with your suggestions.